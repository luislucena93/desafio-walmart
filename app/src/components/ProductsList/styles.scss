@import '../../commons/styleVars';

.productsList {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding-left: 10%;
  padding-right: 10%;
  background-color: $backgroundColor;
  height: calc(100vh - 100px);
  overflow-y: scroll;
}