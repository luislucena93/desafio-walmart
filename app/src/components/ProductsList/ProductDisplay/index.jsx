import React from 'react';
import './styles.scss';
import FormatMoney from '../../../helpers/formatters';

const ProductDisplay = (props) => {
    const {product, addProduct, removeProduct, productQty} = props;

    const getButton = () => {
        if (productQty === 0) {
            return <button onClick={() => addProduct(product)} className={"addButton"}>Agregar</button>
        }
        return (
            <div className={"addRemoveButtons"}>
                <button onClick={() => removeProduct(product)} data-testid={"removeProduct"}>-</button>
                <p>{productQty}</p>
                <button onClick={() => addProduct(product)} data-testid={"addProduct"}>+</button>
            </div>);
    }

    return (
        <div className={"productDisplay"}>
            <div>
                <img className={'thumbnail'} src={`http://${product.image}`}/>
                <div className={"separator"}/>
                <p className={"productDescription"}><span>{product.brand}</span> {product.description}</p>
                <p className={"price"}>${FormatMoney(product.price)}</p>
            </div>
            {getButton()}
        </div>
    )
}

export default ProductDisplay;