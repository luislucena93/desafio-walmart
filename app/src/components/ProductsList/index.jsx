import React, {Fragment} from 'react';
import ProductDisplay from './ProductDisplay/index.jsx';
import './styles.scss';

const ProductsList = (props) => {
    const {products, addProduct, removeProduct, selectedProducts} = props;
    return (
        <Fragment>
            <div className={"productsList"}>
                {products.map((product) => <ProductDisplay key={product._id} product={product} addProduct={addProduct}
                                                           removeProduct={removeProduct}
                                                           productQty={selectedProducts[product.id] ? selectedProducts[product.id].count : 0}/>)}
            </div>
        </Fragment>
    )
};

export default ProductsList;