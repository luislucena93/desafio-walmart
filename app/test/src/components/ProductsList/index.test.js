import React from 'react';
import {render} from '@testing-library/react';
import ProductsList from "../../../../../app/src/components/ProductsList/index";

describe('ProductsList component', () => {

    let mockAddProduct;
    let mockRemoveProduct;
    const testProducts = [{
        _id: 'testId',
        id: 'testId',
        brand: 'testBrand',
        description: 'testDescription',
        image: 'testImage',
        price: 1000
    }];
    const selectedProducts = {testId: {count: 2, info: {...testProducts[0]}}};

    beforeEach(() => {
        mockAddProduct = jest.fn();
        mockRemoveProduct = jest.fn();
    });

    it('should render ProductsList correctly for empty list', () => {
        const {container} = render(<ProductsList selectedProducts={{}} products={[]} addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).toMatchSnapshot();
    });

    it('should render ProductsList correctly for list with no selected products', () => {
        const {container} = render(<ProductsList selectedProducts={{}} products={testProducts}
                                                 addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).toMatchSnapshot();
    });

    it('should render ProductsList correctly for list with selected products', () => {
        const {container} = render(<ProductsList selectedProducts={selectedProducts} products={testProducts}
                                                 addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).toMatchSnapshot();
    });
});