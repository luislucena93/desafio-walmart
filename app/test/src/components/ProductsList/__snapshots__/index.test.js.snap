// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList component should render ProductsList correctly for empty list 1`] = `
<div>
  <div
    class="productsList"
  />
</div>
`;

exports[`ProductsList component should render ProductsList correctly for list with no selected products 1`] = `
<div>
  <div
    class="productsList"
  >
    <div
      class="productDisplay"
    >
      <div>
        <img
          class="thumbnail"
          src="http://testImage"
        />
        <div
          class="separator"
        />
        <p
          class="productDescription"
        >
          <span>
            testBrand
          </span>
           
          testDescription
        </p>
        <p
          class="price"
        >
          $
          1.000
        </p>
      </div>
      <button
        class="addButton"
      >
        Agregar
      </button>
    </div>
  </div>
</div>
`;

exports[`ProductsList component should render ProductsList correctly for list with selected products 1`] = `
<div>
  <div
    class="productsList"
  >
    <div
      class="productDisplay"
    >
      <div>
        <img
          class="thumbnail"
          src="http://testImage"
        />
        <div
          class="separator"
        />
        <p
          class="productDescription"
        >
          <span>
            testBrand
          </span>
           
          testDescription
        </p>
        <p
          class="price"
        >
          $
          1.000
        </p>
      </div>
      <div
        class="addRemoveButtons"
      >
        <button
          data-testid="removeProduct"
        >
          -
        </button>
        <p>
          2
        </p>
        <button
          data-testid="addProduct"
        >
          +
        </button>
      </div>
    </div>
  </div>
</div>
`;
